### ISSUE TITLE
[question] aliasing rules with DropGuard

### ISSUE THREAD

**Issue by WaffleLapkin** [rust-lang/miri#1246]
Recently, [I've discovered that `miri test` errors on my lib][issue] because of aliasing unique refs.

I was trying to hold `*mut _` into the first element of an array and mutate the same array through `index_mut`. MRE could look like this:
```rust
let mut arr: [MaybeUninit<i32>; 1] = [MaybeUninit::uninit()];
let array_base_ptr = arr.as_mut_ptr() as *mut i32;
let mut panic_guard = DropGuard {
    array_base_ptr,
    initialized_count: 0,
};
arr[0] = MaybeUninit::new(1);
panic_guard.initialized_count += 1;
```
([playground][mre])

[mre]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=e09dae94a21343ac37bc12d7f34d5735

<details><summary>miri error</summary>
<p>

```
error: Miri evaluation error: trying to reborrow for Unique, but parent tag <untagged> does not have an appropriate item in the borrow stack
    --> /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/slice/mod.rs:5709:5
     |
5709 |     &mut *ptr::slice_from_raw_parts_mut(data, len)
     |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ trying to reborrow for Unique, but parent tag <untagged> does not have an appropriate item in the borrow stack
     |
note: inside call to `std::slice::from_raw_parts_mut::<i32>` at src/main.rs:22:21
    --> src/main.rs:22:21
     |
22   |                     slice::from_raw_parts_mut(self.array_base_ptr, self.initialized_count);
     |                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: inside call to `<miri_fail::DropGuard<i32> as std::ops::Drop>::drop` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libcore/ptr/mod.rs:177:1
note: inside call to `std::intrinsics::drop_in_place::<miri_fail::DropGuard<i32>> - shim(Some(miri_fail::DropGuard<i32>))` at src/main.rs:40:1
    --> src/main.rs:40:1
     |
40   | }
     | ^
note: inside call to `miri_fail` at src/main.rs:4:5
    --> src/main.rs:4:5
     |
4    |     miri_fail();
     |     ^^^^^^^^^^^
     = note: inside call to `main` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:34
     = note: inside call to closure at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:73
     = note: inside call to closure at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/sys_common/backtrace.rs:130:5
     = note: inside call to `std::sys_common::backtrace::__rust_begin_short_backtrace::<[closure@DefId(1:6041 ~ std[9366]::rt[0]::lang_start_internal[0]::{{closure}}[0]::{{closure}}[0]) 0:&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:52:13
     = note: inside call to closure at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:331:40
     = note: inside call to `std::panicking::r#try::do_call::<[closure@DefId(1:6040 ~ std[9366]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panicking.rs:274:15
     = note: inside call to `std::panicking::r#try::<i32, [closure@DefId(1:6040 ~ std[9366]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe]>` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/panic.rs:394:14
     = note: inside call to `std::panic::catch_unwind::<[closure@DefId(1:6040 ~ std[9366]::rt[0]::lang_start_internal[0]::{{closure}}[0]) 0:&&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe], i32>` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:51:25
     = note: inside call to `std::rt::lang_start_internal` at /playground/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/src/libstd/rt.rs:67:5
     = note: inside call to `std::rt::lang_start::<()>`
```

</p>
</details>

`DropGuard` is needed to drop initialized elements of `arr` if closure (not present in this MRE) panics/fails. `array_base_ptr` is used only in `DropGuard::drop` so I thought that it's ok.

I've used `ptr.add().write()` instead of `arr[] =` and now everything just fine:
```diff
- arr[0] = MaybeUninit::new(1);
+ unsafe { array_base_ptr.add(0).write(1); }
```
([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ab1cef218764b4ea515c689a9926d8eb))

[issue]: https://github.com/WaffleLapkin/arraylib/issues/5

The question is - does my original code really cause UB or this is a miri bug? I have not enogh expirience in unsafe code to understand aliasing rules here :sweat_smile:

**Comment by WaffleLapkin** [comment]
Oh, I forgot to mention: miri errors only on drop of `DropGuard` or more precisely on `slice::from_raw_parts_mut(self.array_base_ptr, self.initialized_count)` call. But at this point, temporary ref created by `index_mut` shouldn't exist, so there shouldn't be any errors, should they?

**Comment by bjorn3** [comment]
This is not allowed by [stacked borrows](https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md), which is a **proposal** for a memory model for rust. It basically requires the normal alias rules that the borrow checker enforces, but at runtime instead of compile time. This allows for certain useful optimizations that would otherwise not #be possible.

**Comment by bjorn3** [comment]
The error happens during the drop of `DropGuard` because at that point the pointer is used again after being invalidated by the temporary ref created by `index_mut`.

**Comment by bjorn3** [comment]
https://www.ralfj.de/blog/2018/11/16/stacked-borrows-implementation.html is an introduction to an older version of stacked borrows. It does still explain the rationale behind stacked borrows and the basic idea is the same.

**Comment by WaffleLapkin** [comment]
@bjorn3 thanks for the explanation!

**Comment by RalfJung** [comment]
On top of what @bjorn3, for your particular case I highly doubt whatever the final rules will be will permit this code. After all, you have a raw pointer and a mutable reference that alias, and you are using both of them in an interleaved fashion. Thus you are violating the principle that mutable references must not alias anything.

> I was trying to hold *mut _ into the first element of an array and mutate the same array through index_mut.

One alternative here is to use raw pointers in your normal code as well.
The other alternative is to create the drop-guard pointer, and then create the pointer for the body *from that pointer*. Then you don't even need unsafe code.
```rust
fn foo(x: &mut [MaybeUninit<i32>]) {
  let guard = DropGuard(x);
  let my_ptr = guard.0;
  // Now you can use my_ptr, and guard's drop also has access to x.
}
```
This approach basically re-structures your code a bit to explicitly let the compiler know that `my_ptr` and the ptr in `guard` are allowed to alias.

**Comment by WaffleLapkin** [comment]
> One alternative here is to use raw pointers in your normal code as well.

That's how I've fixed the issue, yeah.

> The other alternative is to create the drop-guard pointer, and then create the pointer for the body from that pointer. Then you don't even need unsafe code.

Wow, that makes code a lot more clear and removes a lot of `unsafe`s! Thanks, I'll use this solution :heart: 

If I remember correctly I haven't done this in the first place because I've tried to move the array to the `DropGuard`, but then I couldn't move it out because `Drop` is implemented for `DropGuard` (this error is annoying). And then my next idea was `*mut` :sweat_smile:
